### PR TITLE
[iOS] Unexport rust symbols and add new Obj-C wrapper for adblock engine

### PR DESCRIPTION
**Pull request by kylehickinson**
Resolves https://github.com/brave/brave-browser/issues/22208

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by Brandon-T — APPROVED** [review]

**Review by antonok-edm — APPROVED** [review]
Changes under `components` look good to me

File: ios/browser/api/brave_shields/adblock_engine.mm [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  bool did_match_rule = false;
+  bool did_match_exception = false;
+  bool did_match_important = false;
+  std::string redirect = "";
+  adblock_engine->matches(
+      base::SysNSStringToUTF8(url), base::SysNSStringToUTF8(host),
+      base::SysNSStringToUTF8(tabHost), isThirdParty,
+      base::SysNSStringToUTF8(resourceType), &did_match_rule,
+      &did_match_exception, &did_match_important, &redirect);
+  auto result = [[AdblockEngineMatchResult alloc] init];
```
Comment (bridiver) [inline ios/browser/api/brave_shields/adblock_engine.mm:63]: @antonok-edm can make this a mojo type so we can pass it through without copying all the fields?

File: ios/browser/api/brave_shields/adblock_engine.mm [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+#endif
+
+@interface AdblockEngineMatchResult ()
+@property(nonatomic, readwrite) bool didMatchRule;
+@property(nonatomic, readwrite) bool didMatchException;
+@property(nonatomic, readwrite) bool didMatchImportant;
+@property(nonatomic, readwrite, copy) NSString* redirect;
+@end
+
+@implementation AdblockEngineMatchResult
```
Comment (bridiver) [inline ios/browser/api/brave_shields/adblock_engine.mm:24]: can you add thread checks to make sure this class is always called on the same thread/sequence? The adblock engine is not thread safe

File: ios/browser/api/brave_shields/adblock_engine.mm (hunk 1 above)
Comment (kylehickinson) [reply ios/browser/api/brave_shields/adblock_engine.mm:63]: This is mostly just to make the API nicer on Swift side, which doesn't use `inout` arguments as often

File: ios/browser/api/brave_shields/adblock_engine.mm (hunk 2 above)
Comment (kylehickinson) [reply ios/browser/api/brave_shields/adblock_engine.mm:24]: As mentioned in DM, iOS side has the adblock engine wrapped in a serial GCD queues which is incompatible with Chromium's sequenced thread checker.  Would be better to get this part added to the `adblock::Engine` class itself and then remove the individual thread-safety implementations in both brave-core and brave-ios

**Comment by kylehickinson** [comment]
iOS code to support this change will live here: https://github.com/brave/brave-ios/compare/brave-core-1.39.x...adblock-engine-refactor

**Pull request merged**